src/python/pants/python/python_setup.py

```diff
@@ -115,6 +116,17 @@ def register_options(cls, register):
             fingerprint=True,
             help="The maximum number of concurrent jobs to resolve wheels with.",
         )
+        register(
+            "--thirdparty-modules-mapping",
+            type=dict,
+            advanced=True,
+            fromfile=True,
+            help=(
+                "A mapping of your third party requirements to a list of the modules they expose. "
+                "For example, {'ansicolors': ['colors']}. This is used for dependency inference so "
+                "that Pants can understand your import statements."
```

Comment:
Would it help to explicitly mention loading from a file like JSON?

Reply:
Should this maybe be an argument on `python_requirement_library` targets instead? It also feels like it could default to the name of the requirement.

Reply:
I don’t think so. This wouldn’t work with requirements.txt and python_requirements()

Reply:
A requirements.txt is read by the python_requirements() macro and turned into a list of PythonRequirementLibraries, so this should definitely work:
https://github.com/pantsbuild/pants/blob/992a4ca82863be15efa48059abb7ad2f0231c599/src/python/pants/backend/python/python_requirements.py#L67-L76

Reply:
Agreed on how `python_requirements` works. But how would you express that `setuptools==44.0` in a `requirements.txt` file maps to the symbol `pkg_resources`?

Generally, a requirement name == the exported module, but this is not always true. `packaging`, `ansicolors`, and `setuptools` are all examples.

It is also possible for a requirement to have multiple top-level modules, e.g. `pants` vs. `pants_test`.

Reply:
Stu's suggestion was to default. So for all requirements that default, you use requirements.txt. For those that don't, you write them down in the BUILD file that declares python_requirements():
```python
python_requirements()

python_requirement_library(
  name='setuptools',
  requirements=[
    python_requirement(
      'setuptools==44.0',
      top_level=['pkg_resources', 'setuptools'],
    )
  ],
)

...
```

Reply:
Defaulting might be OK (we can check in our data how often that is correct), but overriding in BUILD files seems laborious, as it adds a lot of boilerplate. 

A JSON file, OTOH, would be easy to maintain and even autogenerate.

Reply:
> Stu's suggestion was to default. So for all requirements that default, you use requirements.txt. For those that don't, you write them down in the BUILD file that declares python_requirements():

Or make it an argument to the `python_requirements` macro. But yea.

Reply:
> It's also a new concept to learn

Well, it's a new concept no matter what. It seems very likely that the default won't work for at least one 3rd party requirement, so users will need to learn how to specify the mapping.

More generally, they will need to learn about dep inference / buildgen.

> Or make it an argument to the python_requirements macro. But yea.

My concern here is how decentralized the mapping becomes. You need to scan both your `BUILD` for the `requirements.txt` and any BUILD files that might define an inlined `python_requirement_library`.

I think it's better to have it in one central, unambiguous place. 

> thinking about the documentation for this

The plan is to add a new section to https://pants.readme.io/docs/python-third-party-dependencies called "Teach Pants how to understand your import statements (recommended)", underneath the lockfile section. The JSON file would be recommended to live right next to `requirements.txt`.

I'm not very concerned about being able to teach that quickly and concisely.

Also, reminder that this will usually only need to be done once by a codebase admin. Then when new deps are added, add a mapping, which is easy to figure out based on the other entries.

Reply:
> My concern here is how decentralized the mapping becomes. You need to scan both your BUILD for the requirements.txt and any BUILD files that might define an inlined python_requirement_library.

This is an existing concern: there are already two places. What we would be adding is a third concept.

> My concern here is how decentralized the mapping becomes.
> I think it's better to have it in one central, unambiguous place.

This isn't one unambiguous place... it's two. `3rdparty/**` and somewhere else (potentially). Targets are decentralized because it allows code and config to live together... defining all of your targets in one file wouldn't be great either.

> Also, reminder that this will usually only need to be done once by a codebase admin. Then when new deps are added, add a mapping, which is easy to figure out based on the other entries.

Repo admins aren't the ones adding deps. So that means someone adding a dep needs to know to edit in two places.

Reply:
This ended up being a very good idea. Thanks for the suggestion!